rfcs/2021-05-17-7469-load-balancing-aggregators.md

```diff
@@ -0,0 +1,99 @@
+* A solution outside of Vector itself ensures that users can reliably adopt Vector as an aggregator without replacing their existing infrastructure.
+* Most organizations are likely familiar with operating _some_ class of reverse proxy.
+* Dedicated reverse proxy can be specialized and optimized for its task, and the same can be said for Vector itself.
+
+## Prior Art
+
+* [Logstash: Scaling TCP, UDP, and HTTP](https://www.elastic.co/guide/en/logstash/current/deploying-and-scaling.html#_tcp_udp_and_http_protocols)
+* [Fluentd: Aggregator behind Network Load Balancer](https://aws.amazon.com/blogs/compute/building-a-scalable-log-solution-aggregator-with-aws-fargate-fluentd-and-amazon-kinesis-data-firehose/)
+
+## Drawbacks
+
+* The team will need to maintain a configuration for a third-party application, as well as ensuring the application is kept up-to-date and free of any reported vulnerabilities.
+* We will also need to add the reverse proxy to new or existing integration tests to ensure there are no regressions with our provided configuration and proxy version.
+* Our deployment will be more complex and require an additional application for end users. This can create more misdirection while debugging and additional operational burden.
+
+## Alternatives
+
+### Do Nothing
+
+The Vector aggregator can currently function as a single instance and be scaled vertically rather than horizontally. While this reduces complexity, it causes Vector to be a single point of failure and introduces an upper limit for throughput.
+
+### Only Client-side Load Balancing
+
+The library powering the v2 Vector `sink`/`source` does provide the capabilities to do client-side load balancing, however that just covers a single `sink` to `source` pairing. For certain clients like Beats and Logstash we could implement an Elasticsearch compatible API and allow those clients to use their native load balancing and integrations, this would generally be per source and not available for all sources.
+
+### Require a Service Mesh
+
+Users already leveraging a service mesh could offload the load balancing to the mesh, however, requiring a service mesh to run and scale Vector aggregators horizontally is a large barrier to adoption.
+
+### Distributed hashring
+
+Project like Thanos and Loki have used hashrings to enable multi-tenancy, we could likely do something similar to ensure events are forwarded to the correct aggregator. I don't think anyone wants to turn Vector into a distributed system though.
+
+## Outstanding Questions
+
+* [ ] Which reverse proxy to use? HAProxy, NGINX, Envoy, Traefik, etc. It should be widely used, battle-tested, support most/all protocols Vector uses, and preferably well understood by mutliple members of our team.
+* [ ] Should built-in load balancing capabilities be explored (where possible)? Internal load balancing options would simplify operations for end users who are all-in on Vector.
```

Comment:
I'd be interested in hearing from existing customers here. cc @zsherman

Reply:
I think we'll want this as a generic feature of all sinks, but it should not be our default recommendation for customers wanting to load balance Vector as an aggregator.

Reply:
One of the nice benefits of relying on an external reverse proxy is not needing to bundle this functionality into vector. "All sinks" is a pretty wide space already, wider in the future. But, I can imagine a customer that doesn't want any additional deployed components and needs vector to internalize this work. I'm curious if we know of any today.

Reply:
I suggested this to @jszwedko as well - as a way to accomplish the "logstash" source, mimicking the ES API Logstash would write to would allow us to leverage the client side features of Beats/Logstash. It may be a benefit for individual components, but definitely not the general approach we can take _right now_

Reply:
It would be good to gauge how much interest there is in "internal" load balancing over this proposed "external" solution.

Reply:
As a customer and an operator of vector in production on k8s i would love vector agents to use k8s service as a load balancer to send data to vector aggregator which you might call it the "external" solution.